atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/compact/BackgroundCompactor.java

```diff
@@ -49,40 +50,57 @@
     private final TransactionManager transactionManager;
     private final KeyValueService keyValueService;
     private final LockService lockService;
-    private final Supplier<CompactorConfig> config;
+    private final Supplier<Boolean> inSafeHours;
 
     private Thread daemon;
 
-    public static void createAndRun(TransactionManager transactionManager,
+    public static Optional<BackgroundCompactor> createAndRun(TransactionManager transactionManager,
             KeyValueService keyValueService,
             LockService lockService,
-            Supplier<CompactorConfig> compactorConfigSupplier) {
+            Supplier<Boolean> inSafeHours) {
+        if (!keyValueService.shouldManuallyCompact()) {
+            return Optional.empty();
+        }
+
         BackgroundCompactor backgroundCompactor = new BackgroundCompactor(transactionManager,
                 keyValueService,
                 lockService,
-                compactorConfigSupplier);
+                inSafeHours);
         backgroundCompactor.runInBackground();
+
+        return Optional.of(backgroundCompactor);
     }
 
-    public BackgroundCompactor(TransactionManager transactionManager,
+    private BackgroundCompactor(TransactionManager transactionManager,
             KeyValueService keyValueService,
             LockService lockService,
-            Supplier<CompactorConfig> config) {
+            Supplier<Boolean> inSafeHours) {
         this.transactionManager = transactionManager;
         this.keyValueService = keyValueService;
         this.lockService = lockService;
-        this.config = config;
+        this.inSafeHours = inSafeHours;
     }
 
-    public synchronized void runInBackground() {
+    @Override
+    public void close() {
+        log.info("Closing BackgroundCompactor");
+        daemon.interrupt();
+        try {
+            daemon.join();
+            daemon = null;
+        } catch (InterruptedException e) {
+            throw Throwables.rewrapAndThrowUncheckedException(e);
+        }
+    }
+
+    private synchronized void runInBackground() {
```

Comment:
This is only called in the static factory with a separate instance of the class each time, so I'm not sure why it needs to be `synchronized`.

Reply:
well, this method is still called indirectly by TMs, so we should probably be defensive and keep it synchronised.